PythonSpark/src/Preprocessing.py

```diff
@@ -0,0 +1,127 @@
+    parser.add_argument('--SAKF', dest='service_account_key_filepath',
+                        help='Path to a file with service account credentials', type=str)
+    parser.add_argument('--SAE', dest='service_account_email', help='Email address of a service account', type=str)
+
+    return parser.parse_args()
+
+
+def get_spark_session_for_development(master: str) -> SparkSession:
+    return SparkSession.builder.master(master).appName("Preprocessing app").getOrCreate()
+
+
+def set_up_service_account(service_account_email: str, service_account_key_filepath: str, spark: SparkSession) -> None:
+    spark.conf.set("google.cloud.auth.service.account.enable", "true")
+    spark.conf.set("google.cloud.auth.service.account.email", service_account_email)
+    spark.conf.set("google.cloud.auth.service.account.keyfile", service_account_key_filepath)
+
+
+def get_spark_session_for_production(master: str, service_account_key_filepath: str,
+                                     service_account_email: str) -> SparkSession:
+    spark = SparkSession.builder.master(master).appName("Preprocessing app").getOrCreate()
+    set_up_service_account(service_account_email, service_account_key_filepath, spark)
+    return spark
+
+
+def get_spark_session(args: argparse.Namespace) -> SparkSession:
+    if args.mode == 'dev':
+        return get_spark_session_for_development(args.master)
+    elif args.mode == 'prod':
+        return get_spark_session_for_production(args.master, args.service_account_key_filepath,
+                                                args.service_account_email)
+    else:
+        raise ValueError("Deployment mode can be dev or prod")
+
+
+def load_views(views_filepath: str, spark: SparkSession) -> pyspark.sql.DataFrame:
+    raw_views = spark.read.option("header", "true").csv(views_filepath)
+    parsed_views = raw_views.select(column("user_id"), column("device"), column("ip"),
+                                    to_timestamp(column("ts")).alias("ts"))
+    return parsed_views
+
+
+def repartition_views(views: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
+    return views.repartition(to_date(column("ts")))
+
+
+def process_views(views_filepath: str, repartitioned_views_filepath: str, spark: SparkSession) -> None:
+    logging.info("Started loading views")
+    views = load_views(views_filepath, spark)
+    logging.info("Finished loading views")
+
+    logging.info("Started views repartitioning")
+    repartitioned_views = repartition_views(views)
+    logging.info("Finished views repartitioning")
+
+    logging.info("Started writing views")
+    repartitioned_views.write.parquet(repartitioned_views_filepath)
+    logging.info("Finished writing views")
+
+
+def load_purchases(purchases_filepath: str, spark: SparkSession) -> pyspark.sql.DataFrame:
+    raw_purchases = spark.read.json(purchases_filepath)
+    parse_items = udf(lambda raw_list: [[element[0], int(element[1])] for element in raw_list])
```

Comment:
Use udf only in a case, when no functionality for that. Please read about explode command